### DISCUSSION TITLE
Welcome to python-semver Discussions!

### DISCUSSION THREAD

**Discussion by tomschr** [python-semver/python-semver#329] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by chtaylo3** [comment]
Hi @tomschr , I was wondering if you'd be willing to accept/review a PR for an "extract" class method that takes a string as input and returns a generator (or list) for semver objects? We have some package and vulnerability metadata that we need to parse, so that's the use case. I've been experimenting with [antlr4](https://github.com/antlr/antlr4/) and [this](https://github.com/antlr/grammars-v4/tree/master/semver) grammar as starting point - I think I can put something together that's worth looking at. Provided I can get this up to standard, this could also lay a foundation to help address #241. Please let me know what you think.

**Reply by tomschr** [reply]
Welcome @chtaylo3! :wave: Great to see you here!

> if you'd be willing to accept/review a PR for an "extract" class method that takes a string as input and returns a generator (or list) for semver objects? 

I'm always happy to review PRs to help bringing the project forward. Sure, we can always look at it and see if it fits.

I'm curious. :wink: Could you sketch your use case a bit better? Maybe add some pseudo-code to illustrate your point? Just to get a better picture.

In case it doesn't fit, we can always describe it in the documentation. For example this happened to the pydantic use case. People forget about the documentation, but I try to make it as helpful as possible. :slightly_smiling_face:

Thank you!

**Reply by chtaylo3** [reply]
Thank you for the warm reception @tomschr  :)

The use case is, we have a string that contains all the semver that fix a CVE and we need to parse them all out and then figure out which versions are closest to the vulnerable version currently used.  Think of it as a "least significant change". This comparison is straight forward for major/minor/patch version, but for comparing pre-release, it gets a little trickier. Therefore, having the elements of the pre-release parsed according to the formal grammar of semver can be extremely helpful for comparing different pre-releases.

The psuedo-code is not really straight forward without understanding how antlr4 works, but I'll try to give it a go.  To start, here's the what the method would look like:

```Python
@classmethod
def extract(input_str : str) -> Iterator[semver] :
  lexer = SemanticVersionLexer(input_str)
  parser = SemanticVersionParser(CommonTokenTream(lexer))
  tree = parser.semver()
  walker = ParseTreeWalker()
  return walker.walk(SemanticVersionParserListener(), tree)
```

The lexor code (SemanticVersionLexer) is written for us by anltr based on the grammar we provide and for that grammar, we need to have two modes. The first is a regular mode than looks for the start of a semver and then pushes the lexer into semver mode, discarding all input that isn't the start of a valid semver. The semver mode is for all tokens that are part of semver. The parser generator takes this lexer grammar as input and builds the lexer code for us, in this case python code.  Below is an illustrative example of what I'm working with right now:

```
lexer grammar SemanticVersionLexer;

options { caseInsensitive = true; }

DIGIT2 : [1-9]+  -> pushMode(SEMVER);
Unknown : . -> skip ;

mode SEMVER;
fragment POSITIVE_DIGIT: [1-9];
fragment LETTER: [a-z];

DASH: '-';
PLUS: '+';
DOT: '.';

//most common pre-release "modifiers"
ALPHA: 'alpha';
BETA: 'beta';
RC: 'rc' | 'release' ('-' | '.') 'candidate';
SNAPSHOT: 'snapshot';
PREVIEW: 'p' | 'pre' | 'preview';
DEV: 'dev' | 'devel' | 'development';
MILESTONE: 'mt' | 'milestone';
DAILY: 'daily';
NIGHTLY: 'nightly';
BUILD: 'bld' | 'build';
TEST: 'test';
EXPERIMENTAL: 'experimental';

NUMBER: '0' | POSITIVE_DIGIT+;

IDENTIFIER: LETTER+;

Unknown2 : . -> popMode;
```
Similarly, the parser code is built for us (SemanticVersionParser) from the following parser grammar (again, this is just what I'm working with right now and still needs some improvements). We might want to align this closer to the formal definitions from semver.
 ```
parser grammar SemanticVersionParser;

options { tokenVocab = SemanticVersionLexer; }

tag: ALPHA
    | BETA
    | RC
    | SNAPSHOT
    | PREVIEW
    | DEV
    | MILESTONE
    | DAILY
    | NIGHTLY
    | BUILD
    | TEST
    | EXPERIMENTAL
    ;

build:
      value = NUMBER                                             #BuildNumber
    | DASH                                                       #BuildDash
    | tag sep = (DOT | DASH)? version = (NUMBER  | IDENTIFIER)?  #BuildTagged
    | value = IDENTIFIER                                         #BuildIdentifier
    | left = build (sep = DOT? right = build)+                   #BuildIdentifierExtended
    | left = build (sep = DASH? right = build)+                  #BuildIdentifierDash
    ;

preRelease:
      value = NUMBER                                            #PreReleaseNumber
    | DASH                                                      #PreReleaseDash
    | tag sep = (DOT | DASH)? version = (NUMBER | IDENTIFIER)?  #PreReleaseTagged
    | value = IDENTIFIER                                        #PreReleaseIdentifier
    | left = preRelease (sep = DOT? right = preRelease)+        #PreReleaseIdentifierExtended
    | left = preRelease (sep = DASH? right = preRelease)+       #preReleaseIdentifierDash
    ;

versionCore: major = NUMBER DOT minor = NUMBER (DOT patch = NUMBER)?;

semver: versionCore (DASH preRelease (PLUS build)? | PLUS build)?EOF;
```
Then we extend the parser listener class, via SemanticVerionParserListener, and decorate each entry/exit method from the listener with either a decorator for handling atoms or aggregates. Here is an example of the one for atoms:

```Python
   def _atom(T: Type[Type] = None):
        
        def _atom_decorator(func):
            
            @wraps(func)
            def _atom_wrapper(*args, **kwargs):
                #TODO: figure out what logging infra to use instead of print this out
                func_name : str = func.__name__
                predicate : str = 'enter' if func_name.startswith('enter') else 'exit'
                func_name = func_name.removeprefix(predicate)
                if T is not None : args[1].values = [T(args[1].getText())]
                print(f'{predicate.title()} {func_name}: {args[1].getText()} : {args[1].values if T is not None else ""}')
                return func(*args, **kwargs)
            return _atom_wrapper
        
        return _atom_decorator
   ``` 

The use of parser generators (like antlr4) instead of regular expressions will also allow much easier implementations of new formal grammars, for example if semver 3 comes out or the formal grammar listed in #241. The downside of course is using a parser generators is a bit more complex than regex.

**Reply by tomschr** [reply]
Thanks you for your detailed answer and helping me to understand it better! Much appreciated. :+1:

I think I get slowly an idea about this topic. Maybe it's due to my cold from which I haven't yet fully recovered, but I'm not completely sure of your suggested change. Am I right that you would like to add this `extract` class method to the `Version` class and the result from the Antlr grammar output (the parser)?

That would be an interesting idea. Maybe you can answer my concerns:

* Simplicity & dependencies
  The semver library (and specifically the `Version` class) is quite simple. It doesn't have any dependencies and I would like to keep it that way. Maybe I got it wrong, but adding some dependencies of antlr (even if it's generated code by that) sounds a bit... too much to me?

* Code generations
  If I'm not mistaken, as antlr is a parser generator and create Python code from your grammar. That sounds quite awesome. But for such a small project it sounds a bit of an overkill and the code you get is probably not very pythonic. Sure, nobody wants to write a parser manually, but I fear this code would be bigger than the whole class itself.
  How much Python code lines do we talk about?

* Semver specification
  Apart from the previous concerns semver itself doesn't specifies any tags (be it "alpha", "snapshots" etc.) as literals. If you add them to your grammar, you give them a special meaning that the semver specification itself doesn't define. For semver it's just a "prerelease" string. Adding those to the class sounds a bit of an extension to semver.

I'm a big fan of making the project or the class as flexible and adaptable as possible to different use cases. One way would be, for example, to enable others to derive the class and to implement something that the base class cannot provide. This is how Pydantic does it and it is described in the documentation.

The other way could be to improve issue #241: I've played around a bit and created PR #367 some time ago. It's certainly not perfect, but it tries to implement the tilde and caret operators. It's not all fleshed out, as there are currently some design issues to be solve. But it's a start. What the PR currently lacks is support for prereleases. Maybe the base class could provide some way to parse the prerelease tags to support more fancy comparisions? Would that help your use case if that issue would be supported?

I'm not sure if this helps or answers your questions. Perhaps this current post touches more the philosophical aspect than the mere technical one. :slightly_smiling_face: 

Would like to get to know your point of view and how we can solve it together.

**Reply by chtaylo3** [reply]
Hope you feel better soon!

I agree with your statement about the philosophical aspects, but I think there may be a way to address both.

Q: Am I right that you would like to add this extract class method to the Version class and the result from the Antlr grammar output (the parser)?
A: Yes. It would also be possible to relegate this extract capability to an optional component of the package so that users still retain the ability to download just the smaller core capability.

Q: Concerns regarding Simplicity & dependencies
A: Yes, these are all accurate and valid concerns. There is a runtime dependency as well as the code generated by antlr. Making this an optional component of the package should address these concerns.

Q: Concerns regarding code generation
A: Yes, antlr or lex/yacc or whatever are parser generators that build a parser based on the grammar provided. In addition to the code generated on our behalf, there is an additional runtime dependency. One of the reasons why I gravitated towards antlr4 is this rewrite was done purposefully to make it easier for people that want to parse things other than a full grammar for a compiler (see the listener/visitor pattern) and the output is significantly more pythonic than v3 was. The lexer is about ~150-200 loc, the parser is about ~1 KLOC (it grows or shrinks depending on the number of rules in the parser grammar). The listener that we inherit from is about ~100 LOC of just methods that 'pass' so they're not functional lines of code. The inherited listener (where we write our code) will have ~2 or 3 decorators of about 50-200 LOC each which then need to be applied on the various methods. There is a dependency on the antlrv4 python runtime. I don't know how many LOC that is, but I can find out if you need. In terms of runtime, I would _assume_ it's around the same number of LOC _executed_ as a regex engine although the re module is not installed with the package because it's in stdlib. In summary, yes - it would noticeably increase the size, but I think we can mitigate that by having this functionality provided in some sort of optional install for the package. This might mean it doesn't go directly into the version class as a class method, but I think the tradeoff would be worth it.

Q: Concerns regarding semver specification.
A: I'm more than happy to revert that back to the spec and remove these additional tags. There's always a tradeoff between the formal spec and how people use it.

Q:  Maybe the base class could provide some way to parse the prerelease tags to support more fancy comparisions? Would that help your use case if that issue would be supported?
A: Partially. We're getting comma separated semvers in a string which tells people what they should upgrade to. We could parse that by hand if needed but I thought a generalized extract method might be more useful to others. We're also getting names of libraries that we need to parse the version out from as well.  For example, "my_lib_compat_1.2.3_4.5.6-rc5.7-3.8.23.jar", where 4.5.6-rc5.7-3.8.23 is the semver. My only concern is that this can get complicated very quickly. I'd be happy to think on that and see if I think I can submit a PR for this without using a parser generator.

My concern is that if you try and _fully_ implement #241, you're going to run into problems trying to do this by hand compared to using an existing parser generator. That's why I was thinking about a parser for an extract capability because it could also lay the foundation to address both problems easily.

**Reply by tomschr** [reply]
Thanks for your answers! :+1: 

> Hope you feel better soon!

Thanks, still dealing with some coughing and I'm slow to respond. But mostly I'm fine. :slightly_smiling_face:

I've tried to compile your examples with antlr4, but I wasn't successful:

```
$ antlr4  SemanticVersionParser.g4
warning(125): SemanticVersionParser.g4:5:5: implicit definition of token ALPHA in parser
warning(125): SemanticVersionParser.g4:6:5: implicit definition of token BETA in parser
warning(125): SemanticVersionParser.g4:7:5: implicit definition of token RC in parser

... similar warning messages pruned ...
```

There was no Python code created. Obviously, I did something wrong here. :laughing: 



> In summary, yes - it would noticeably increase the size, but I think we can mitigate that by having this functionality provided in some sort of optional install for the package.

Even if we make it "optional", we still have to build, test, document, and maintain it. If it's there, people will use it.


> [... Spec...] I'm more than happy to revert that back to the spec and remove these additional tags. There's always a tradeoff between the formal spec and how people use it.

Well, specifications are there for a reason. This project tries to follow the semver specification as best as it can.

You can try to convince the semver people. Perhaps they will add it in one of their next version. However, I had the impression that they wanted to keep their specification lean and easy. 

> [...] but I thought a generalized extract method might be more useful to others.

Yes, maybe. I'm always driven by this simple question: What should be supported/integrated by semver (or more specifically, the Version class) and what not?

I think, the obvious core purpose is to do one task well: managing semver versions. This involves converting to and from strings, tuples, dicts, raising versions, and comparing versions to each other.

What is outside of those core tasks is nothing that semver will do. For example, reading versions from a file or connecting to a database to extract versions from there. This is something the user has to take care of.

What I'm trying to say is, if it follows its core purpose and is beneficial to a broader audience, then it's very likely it will get incorporated. If it's something that is only beneficial to parts of our users, it could be probably be integrated into the documentation.

The question is, would such an extract method be useful for others? How often does it occur and is needed? For the ordinary users who are only interested in comparing versions is probably not very interesting. You have a special use case and I assume, only a fraction of semver's users would need it. But I could be wrong.


> My concern is that if you try and fully implement #241, you're going to run into problems trying to do this by hand compared to using an existing parser generator.

Oh, quite likely! I wrote so many wrong lines that I can't count them. :laughing:

So yes, that can happen. But for this we have the test suite to ensure code is heavily tested and works as advertised. Of course, it's not a 100% guarantee. But you don't have a guarantee either that your generated parser with ~1200 LOC doesn't have bugs. It's a heavy piece. :wink: 

semver's library has got the same ~1200 LOC (wheres the core with `version.py` is appr. half the size). Adding another ~1200 LOC just for the parsing the same that already exists sounds a bit, uhmn, overkill to me and a bad trade-off.

At the moment, the Version class uses a regex for its "parsing". It's simple, effective, and fast. I don't see yet the need to add an additional dedicated parser that is a) quite heavy, b) implements what's already there, and c) adds additional dependencies.

Sooo... sorry, ATM I'm not really convinced this would be something useful for this project. But I could be wrong, so let me think and digest it for some more days.

But if you have ideas how to improve #241 I'm more than happy!

**Reply by chtaylo3** [reply]
You need to do the lexer first so the parser has something to work from.  you also probably want to pass along the target language and the version of the runtime you're compiling for.  So for example

```
poetry add antlr4-python-runtime:4.13.1
poetry shell
antlr4 -v 4.13.1 -Dlanguage=Python3 SemanticVersionLexer.g4
antlr4 -v 4.13.1 -Dlanguage=Python3 SemanticVersionParser.g4
//whatever code editor
```
> Sooo... sorry, ATM I'm not really convinced this would be something useful for this project.

No worries at all! One of the things we were hoping to get from this is a tokenized/parsed pre-release for comparison purposes so that we can later do some analysis on the release velocity and such. 

> But if you have ideas how to improve https://github.com/python-semver/python-semver/issues/241 I'm more than happy!

Unfortunately, I would use antlr4 to generate a parser from the formal grammar so that we don't have to reinvent things. I'm a bit lazy like that ;)

**Reply by tomschr** [reply]
Thanks for your understanding. :heart:

I'm still curious about antlr and would try it out a bit, but for some reason this doesn't work as expected. I've tried it in a subdirectory (without `pyproject.toml`, so I had to use `poetry init` first). I'm stuck at the first step:

```
$ poetry add antlr4-python-runtime:4.13.1

Could not find a matching version of package antlr4-python-runtime

$ poetry add antlr4-python-runtime==4.13.1

Could not find a matching version of package antlr4-python-runtime
```

I normally work with `pyenv` and I've already installed the above lib (although with `pip`). If I skip the poetry step, I see these two `.g4` files, but when I try to run the first one, I get this:

```
$ antlr4 -v 4.13.1 -Dlanguage=Python3 SemanticVersionLexer.g4
error(2):  unknown command-line option -v
error(7):  cannot find or open file: 4.13.1
error(50): SemanticVersionLexer.g4:35:9: syntax error: mismatched input ',' expecting COLON while matching a lexer rule
error(50): SemanticVersionLexer.g4:35:65: syntax error: mismatched input ')' expecting COLON while matching a lexer rule
```

Removing `-v 4.13.1` doesn't help either. Any idea?


> > ideas how to improve #241 
> Unfortunately, I would use antlr4 to generate a parser from the formal grammar so that we don't have to reinvent things. I'm a bit lazy like that ;)

That's totally fine. I was more referring to the overall design than adding code. :slightly_smiling_face:
